### PR TITLE
[lockfile] Add lockfile and use search endpoint to resolve versions

### PR DESCRIPTION
**Pull request by mikeland86**
## Summary

This introduces a simple lock file and uses the search service to resolve versions. The lock file is created/updated whenever a package is added/removed. If a package is manually added to devbox.json the lockfile is updated on run/shell/install/shellenv.

Packages that don't have `@` are never added to lockfile.

This is behind a feature flag.

## How was it tested?

```bash
devbox add python@3.9
devbox run "python --version"
devbox rm python@3.9
```

* Inspected nix profile to ensure packages are added and removed as expected.
* Inspected lock file to ensure contents are as expected.

### CODE REVIEW

**Comment by mikeland86** [comment]
@loreto @gcurtis this is ready to review.

**Review by loreto — APPROVED** [review]
LGTM

**Comment by loreto** [comment]
It looks like there's some conflicts you need to resolve

**Review by gcurtis — APPROVED** [review]

File: internal/nix/input.go [hunk 1]
```diff
@@ -57,19 +71,28 @@ func (i *Input) Name() string {
...
-	u := i.URL // get copy
-	u.Fragment = ""
-	// This will produce urls with extra slashes after the scheme, but that's ok
-	return u.String()
+func (i *Input) URLForInput() string {
+	if i.IsDevboxPackage() {
+		resolved, err := i.lockfile.Resolve(i.String())
+		if err != nil {
+			panic(err)
+			// TODO(landau): handle error
```
Comment (savil) [inline internal/nix/input.go:79]: @mikeland86 gentle reminder to handle this panic's TODO (in case it slips through the cracks)

**Review by savil — COMMENTED** [review]
catching up on this PR so I can review #964...

**Pull request merged**